stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -2981,13 +2997,32 @@ internal struct _Native${Self}Buffer<${TypeParameters}> {
   }
 
 %if Self == 'Set':
-  @_transparent
-  internal func initializeKey(_ k: Key, at i: Int) {
-    _sanityCheck(!isInitializedEntry(at: i))
+  internal func initializeKey(_ k: Key, hash h: Int, at i: Int) {
     defer { _fixLifetime(self) }
-
+    initializeKeyHelper(k, hash: h, at: i)
+  }
+  
+  // Never call this function directly, it is called by initializeKey
+  // and performs recursion, we use indirection to guarantee tail call
+  // optimization
+  internal func initializeKeyHelper(_ k: Key, hash h: Int, at i: Int) {
```

Comment:
This assumption is unstable.  Would it be especially burdensome to rewrite this with loops instead?

As for the visibility issue (separate from TCO), does making this a nested function still cause TCO to kick in?